### ISSUE TITLE
Buffer Leak in LzfIntegrationTest

### ISSUE THREAD

**Issue by danbev** [netty/netty#3312]
CI [reported](http://clinker.netty.io/jenkins/job/netty-pullreqs/457/consoleFull) a buffer leak which looks like it is coming from ```LzfIntegrationTest```. 
The section relevant to the leak is the following:
```
14:57:42 14:57:42.272 [main] ERROR io.netty.util.ResourceLeakDetector - LEAK: ByteBuf.release() was not called before it's garbage-collected.
14:57:42 Recent access records: 2
14:57:42 #2:
14:57:42 	Hint: 'EmbeddedChannel$LastInboundHandler#0' will handle the message from this point.
14:57:42 	io.netty.buffer.CompositeByteBuf.touch(CompositeByteBuf.java:1614)
14:57:42 	io.netty.buffer.CompositeByteBuf.touch(CompositeByteBuf.java:40)
14:57:42 	io.netty.buffer.AbstractDerivedByteBuf.touch(AbstractDerivedByteBuf.java:56)
14:57:42 	io.netty.buffer.AbstractDerivedByteBuf.touch(AbstractDerivedByteBuf.java:25)
14:57:42 	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:156)
14:57:42 	io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:182)
14:57:42 	io.netty.channel.ChannelHandlerInvokerUtil.invokeChannelReadNow(ChannelHandlerInvokerUtil.java:83)
14:57:42 	io.netty.channel.embedded.EmbeddedEventLoop.invokeChannelRead(EmbeddedEventLoop.java:157)
14:57:42 	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:157)
14:57:42 	io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:946)
14:57:42 	io.netty.channel.embedded.EmbeddedChannel.writeInbound(EmbeddedChannel.java:176)
14:57:42 	io.netty.handler.codec.compression.IntegrationTest.testIdentity(IntegrationTest.java:129)
14:57:42 	io.netty.handler.codec.compression.IntegrationTest.testEmpty(IntegrationTest.java:42)
14:57:42 	sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
14:57:42 	sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
14:57:42 	sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
14:57:42 	java.lang.reflect.Method.invoke(Method.java:606)
14:57:42 	org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
14:57:42 	org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
14:57:42 	org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
14:57:42 	org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
14:57:42 	org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
14:57:42 	org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
14:57:42 	org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
14:57:42 	org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
14:57:42 	org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
14:57:42 	org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
14:57:42 	org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
14:57:42 	org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
14:57:42 	org.junit.runners.ParentRunner.run(ParentRunner.java:309)
14:57:42 	org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:264)
14:57:42 	org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153)
14:57:42 	org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:124)
14:57:42 	org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:200)
14:57:42 	org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:153)
14:57:42 	org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:103)
14:57:42 #1:
14:57:42 	Hint: 'LzfDecoder#0' will handle the message from this point.
14:57:42 	io.netty.buffer.CompositeByteBuf.touch(CompositeByteBuf.java:1614)
14:57:42 	io.netty.buffer.CompositeByteBuf.touch(CompositeByteBuf.java:40)
14:57:42 	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:156)
14:57:42 	io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:946)
14:57:42 	io.netty.channel.embedded.EmbeddedChannel.writeInbound(EmbeddedChannel.java:176)
14:57:42 	io.netty.handler.codec.compression.IntegrationTest.testIdentity(IntegrationTest.java:129)
14:57:42 	io.netty.handler.codec.compression.IntegrationTest.testEmpty(IntegrationTest.java:42)
14:57:42 	sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
14:57:42 	sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
14:57:42 	sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
14:57:42 	java.lang.reflect.Method.invoke(Method.java:606)
14:57:42 	org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
14:57:42 	org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
14:57:42 	org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
14:57:42 	org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
14:57:42 	org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
14:57:42 	org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
14:57:42 	org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
14:57:42 	org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
14:57:42 	org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
14:57:42 	org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
14:57:42 	org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
14:57:42 	org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
14:57:42 	org.junit.runners.ParentRunner.run(ParentRunner.java:309)
14:57:42 	org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:264)
14:57:42 	org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153)
14:57:42 	org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:124)
14:57:42 	org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:200)
14:57:42 	org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:153)
14:57:42 	org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:103)
14:57:42 Created at:
14:57:42 	io.netty.buffer.CompositeByteBuf.<init>(CompositeByteBuf.java:60)
14:57:42 	io.netty.buffer.Unpooled.compositeBuffer(Unpooled.java:359)
14:57:42 	io.netty.buffer.Unpooled.compositeBuffer(Unpooled.java:352)
14:57:42 	io.netty.handler.codec.compression.IntegrationTest.testIdentity(IntegrationTest.java:122)
14:57:42 	io.netty.handler.codec.compression.IntegrationTest.testEmpty(IntegrationTest.java:42)
14:57:42 	sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
14:57:42 	sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
14:57:42 	sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
14:57:42 	java.lang.reflect.Method.invoke(Method.java:606)
14:57:42 	org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
14:57:42 	org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
14:57:42 	org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
14:57:42 	org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
14:57:42 	org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
14:57:42 	org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
14:57:42 	org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
14:57:42 	org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
14:57:42 	org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
14:57:42 	org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
14:57:42 	org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
14:57:42 	org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
14:57:42 	org.junit.runners.ParentRunner.run(ParentRunner.java:309)
14:57:42 	org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:264)
14:57:42 	org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153)
14:57:42 	org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:124)
14:57:42 	org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:200)
14:57:42 	org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:153)
14:57:42 	org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:103)
14:57:42 Tests run: 10, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.109 sec - in 
```

**Comment by trustin** [comment]
`LzfDecoder` had a leak when handling an empty frame. Fixed in: 3ebe2ee36978fca5e3e4f8861e81f9b61a2c7549

**Comment by danbev** [comment]
@trustin Awesome, thanks for info and ref to the commit.
